### PR TITLE
feat: Add E2E Tests for Phi-3 and Tuning

### PR DESCRIPTION
**Pull request by ishaansehgal99**
**Reason for Change**:
1. Add E2E tests for Phi-3 and Tuning
2. Uses temporary ACR for uploading e2e results
3. Allows a public preset to be used in private accessmode with private ACR path 
4. Adds Validation check for count > 1 for Tuning

### CODE REVIEW

File: pkg/controllers/workspace_controller.go [hunk 1]
```diff
@@ -492,18 +503,20 @@ func (c *WorkspaceReconciler) ensureService(ctx context.Context, wObj *kaitov1al
...
-		err = resources.CreateResource(ctx, serviceObj, c.Client)
-		if err != nil {
+		if err := resources.CreateResource(ctx, serviceObj, c.Client); err != nil {
 			return err
 		}
 		if model.SupportDistributedInference() {
 			headlessService := resources.GenerateHeadlessServiceManifest(ctx, wObj)
-			err = resources.CreateResource(ctx, headlessService, c.Client)
-			if err != nil {
+			if err := resources.CreateResource(ctx, headlessService, c.Client); err != nil {
```
Comment (ishaansehgal99) [inline pkg/controllers/workspace_controller.go:519]: These code changes ensure workspace service is created for tuning....however we could remove this given tuning has no endpoints...

File: pkg/controllers/workspace_controller.go [hunk 2]
```diff
@@ -471,6 +471,17 @@ func (c *WorkspaceReconciler) ensureNodePlugins(ctx context.Context, wObj *kaito
 	}
 }
 
+// getPresetName returns the preset name from wObj if available
+func getPresetName(wObj *kaitov1alpha1.Workspace) *string {
```
Comment (Fei-Guo) [inline pkg/controllers/workspace_controller.go:475]: "A string has two values in it: pointer to an array, and the string length. When you pass string as an argument, those two values are copied, not the underlying array." 

It is ok to return string type to simplify code, not much overhead.

File: pkg/controllers/workspace_controller.go (hunk 1 above)
Comment (Fei-Guo) [reply pkg/controllers/workspace_controller.go:519]: We should skip the entire ensureService for tuning.

File: api/v1alpha1/workspace_validation.go [hunk 3]
```diff
@@ -45,7 +45,7 @@ func (w *Workspace) Validate(ctx context.Context) (errs *apis.FieldError) {
 		errs = errs.Also(w.validateCreate().ViaField("spec"))
 		if w.Inference != nil {
 			// TODO: Add Adapter Spec Validation - Including DataSource Validation for Adapter
-			errs = errs.Also(w.Resource.validateCreate(*w.Inference).ViaField("resource"),
+			errs = errs.Also(w.Resource.validateCreate(w.Inference, w.Tuning).ViaField("resource"),
 				w.Inference.validateCreate().ViaField("inference"))
 		}
 		if w.Tuning != nil {
```
Comment (Fei-Guo) [inline api/v1alpha1/workspace_validation.go:51]: We need to have separate resource check for tuning.

File: api/v1alpha1/workspace_validation.go [hunk 4]
```diff
@@ -299,16 +299,19 @@ func (r *DataDestination) validateUpdate(old *DataDestination) (errs *apis.Field
...
 }
 
-func (r *ResourceSpec) validateCreate(inference InferenceSpec) (errs *apis.FieldError) {
+func (r *ResourceSpec) validateCreate(inference *InferenceSpec, tuning *TuningSpec) (errs *apis.FieldError) {
 	var presetName string
-	if inference.Preset != nil {
+	if inference != nil {
 		presetName = strings.ToLower(string(inference.Preset.Name))
 	}
+	if tuning != nil && *r.Count > 1 {
```
Comment (Fei-Guo) [inline api/v1alpha1/workspace_validation.go:307]: Let us move this to another method for checking tuning resource.

File: api/v1alpha1/workspace_validation.go [hunk 5]
```diff
@@ -45,7 +45,7 @@ func (w *Workspace) Validate(ctx context.Context) (errs *apis.FieldError) {
 		errs = errs.Also(w.validateCreate().ViaField("spec"))
 		if w.Inference != nil {
```
Comment (Fei-Guo) [inline api/v1alpha1/workspace_validation.go:46]: We need to ensure either one of tunning or inference spec should be set, not both.

File: pkg/inference/preset-inferences.go [hunk 6]
```diff
@@ -96,7 +96,8 @@ func updateTorchParamsForDistributedInference(ctx context.Context, kubeClient cl
 
 func GetInferenceImageInfo(ctx context.Context, workspaceObj *kaitov1alpha1.Workspace, presetObj *model.PresetParam) (string, []corev1.LocalObjectReference) {
 	imagePullSecretRefs := []corev1.LocalObjectReference{}
-	if presetObj.ImageAccessMode == string(kaitov1alpha1.ModelImageAccessModePrivate) {
+	if presetObj.ImageAccessMode == string(kaitov1alpha1.ModelImageAccessModePrivate) ||
+		string(workspaceObj.Inference.Preset.AccessMode) == string(kaitov1alpha1.ModelImageAccessModePrivate) {
```
Comment (Fei-Guo) [inline pkg/inference/preset-inferences.go:100]: Why do we need the extra check?

File: test/e2e/preset_test.go [hunk 7]
```diff
@@ -109,16 +118,76 @@ func createCustomWorkspaceWithPresetCustomMode(imageName string, numOfNode int)
...
 				MatchLabels: map[string]string{"kaito-workspace": "private-preset-e2e-test-custom"},
-			}, nil, "", utils.InferenceModeCustomTemplate, nil, utils.GeneratePodTemplate(uniqueID, namespaceName, imageName, nil))
+			}, nil, "", utils.CustomTemplateAccessMode, nil, utils.GeneratePodTemplate(uniqueID, namespaceName, imageName, nil))
 
 		createAndValidateWorkspace(workspaceObj)
 	})
 	return workspaceObj
 }
 
+func createPhi3WorkspaceWithPresetPrivateMode(registry, registrySecret, imageVersion string, numOfNode int) *kaitov1alpha1.Workspace {
```
Comment (Fei-Guo) [inline test/e2e/preset_test.go:131]: Can we use public image for e2e test?

File: test/e2e/utils/utils.go [hunk 8]
```diff
@@ -36,6 +42,15 @@ func GetEnv(envVar string) string {
 	return env
 }
 
+func GenerateRandomString(n int) string {
```
Comment (Fei-Guo) [inline test/e2e/utils/utils.go:45]: Maybe make it simpler, just a random number between 0 and 1000 and convert to string.

File: api/v1alpha1/workspace_validation.go (hunk 3 above)
Comment (ishaansehgal99) [reply api/v1alpha1/workspace_validation.go:51]: Added

File: pkg/controllers/workspace_controller.go (hunk 2 above)
Comment (ishaansehgal99) [reply pkg/controllers/workspace_controller.go:475]: We return a pointer to perform nil check

File: pkg/inference/preset-inferences.go (hunk 6 above)
Comment (ishaansehgal99) [reply pkg/inference/preset-inferences.go:100]: This check allows public models to run in private mode.

File: api/v1alpha1/workspace_validation.go (hunk 5 above)
Comment (ishaansehgal99) [reply api/v1alpha1/workspace_validation.go:46]: func (w *Workspace) validateCreate() (errs *apis.FieldError) {
	if w.Inference == nil && w.Tuning == nil {
		errs = errs.Also(apis.ErrGeneric("Either Inference or Tuning must be specified, not neither", ""))
	}
	if w.Inference != nil && w.Tuning != nil {
		errs = errs.Also(apis.ErrGeneric("Either Inference or Tuning must be specified, but not both", ""))
	}
	return errs
}

File: pkg/controllers/workspace_controller.go (hunk 1 above)
Comment (ishaansehgal99) [reply pkg/controllers/workspace_controller.go:518]: Resolved

File: presets/models/falcon/model.go [hunk 9]
```diff
@@ -75,7 +75,7 @@ func (*falcon7b) GetTuningParameters() *model.PresetParam {
 		ModelFamilyName:           "Falcon",
 		ImageAccessMode:           string(kaitov1alpha1.ModelImageAccessModePublic),
 		DiskStorageRequirement:    "50Gi",
-		GPUCountRequirement:       "2",
+		GPUCountRequirement:       "1",
```
Comment (bangqipropel) [inline presets/models/falcon/model.go:78]: Just question, why we change it back to 1?

**Review by bangqipropel — APPROVED** [review]
Approved

File: presets/models/falcon/model.go (hunk 9 above)
Comment (ishaansehgal99) [reply presets/models/falcon/model.go:78]: We can run falcon tuning with 1 gpu, its just that its slower, but still possible

**Review by Fei-Guo — APPROVED** [review]

**Pull request merged**